### DISCUSSION TITLE
Layout settings ignored when UI is run in Python

### DISCUSSION THREAD

**Discussion by 2briancox** [alejandroautalan/pygubu-designer#161] [Q&A] [answered]
I have created a custom UI that uses TkScrolledFrame and TkScrolledText and when I use Pygubu Designer to open it (F5) every widget behaves as I want upon resizing the window. 

But when I open the UI in a Python script, the window is stagnant.  When I resize, the window shows white space beyond the initial window size when I make the window larger, and the window simply cuts off anything if I make it smaller than the initial window size. 

I am getting a Python error:
_Failed to set property 'container' on class '<class 'pygubu.widgets.tkscrolledframe.TkScrolledFrame'>'. Error: can't modify -container option after widget is created_

Hopefully when I resolve that error this problem will go away.  Changed the UI file to a txt file for upload.

[GUI.txt](https://github.com/alejandroautalan/pygubu-designer/files/9904887/GUI.txt)

For reference, here is my Python code.   Saved as a txt file.
[CMM_Workflow_Client.txt](https://github.com/alejandroautalan/pygubu-designer/files/9904915/CMM_Workflow_Client.txt)

**Comment by alejandroautalan** [comment, marked as answer]
Hello @2briancox thanks for trying pygubu.

Lets start with this:

> I am getting a Python error: _Failed to set property 'container' on class '<class 'pygubu.widgets.tkscrolledframe.TkScrolledFrame'>'. Error: can't modify -container option after widget is created_

You have set (maybe accidentally) container = True in the frame. This is a special property used to embed an entire toplevel window inside a frame. If you set to empty or False the error will disappear.

![2022-10-31_22-14](https://user-images.githubusercontent.com/3482471/199142377-426f2412-dbdf-48e3-b213-053843d75a09.png)

> I have created a custom UI that uses TkScrolledFrame and TkScrolledText and when I use Pygubu Designer to open it (F5) every widget behaves as I want upon resizing the window.
> 
> But when I open the UI in a Python script, the window is stagnant. When I resize, the window shows white space beyond the initial window size when I make the window larger, and the window simply cuts off anything if I make it smaller than the initial window size.

I found some problems here.  When you use grid you have to try not to leave gaps in the grid. For example here you positioned the frames in rows 5 and 6, so rows 0 to 4 are empty. This can cause problems later when configuring the elasticity of the grid.

![2022-10-31_22-21](https://user-images.githubusercontent.com/3482471/199144191-7f1ce316-8fad-43ee-b08b-25fadba06694.png)

As you are using a frame as the main widget of the UI, you need to configure the stretching of the toplevel:

```python
def __init__(self, master):
    ....
    builder.add_from_file('{}\\Resources\\GUI.ui'.format(self.fullPath))
    self.mainwindow = builder.get_object('frame_1', master)
    self.master.minsize(940, 0)
    # Configure stretching of master.
    self.master.columnconfigure(0, weight=1)
    self.master.rowconfigure(0, weight=1)
    #      
    builder.connect_callbacks(self)
```

> and the window simply cuts off anything if I make it smaller than the initial window size.

This is the normal behaviour of Tk. You can avoid that setting a minimum size for the toplevel.

```python
def __init__(self, master):
    ....
    # Configure stretching of master.
    self.master.columnconfigure(0, weight=1)
    self.master.rowconfigure(0, weight=1)
    #  Configure minimum size
    self.master.geometry("1140x800")
    self.master.minsize(1140, 800)
    self.master.resizable(True, True)
    builder.connect_callbacks(self)
```

Hope this helps you. Let me know if you have more questions.

Regards
Alejandro A.

**Reply by 2briancox** [reply]
Thank you again, for the second time in 3 years, for your excellent explanation to one of my challenges.  That worked perfectly on all accounts.  I owe you a beer.

I have a couple of follow up questions.  When I stretch the window I now get the vertical spacing between rows stretching.  I do not want this.  Why does that happen that way?  I would like those to remain tight together at the defined 7 pady pizel size.

Also, fact I only want 2 particular rows within my labelframes stretching.  Within labelframe2 I only want rows 7 and 9 to stretch (2 tkscrolledtexts).  See attached UI:
[GUI.txt](https://github.com/alejandroautalan/pygubu-designer/files/9913186/GUI.txt)

Lastly, I cannot figure out how to fix the size of the tkscrolledframe_1.  No matter what I do, the height is determined by the window size.  You will see within the appearance, I have the height set to only 200.  But nothing changes with the height regardless of how I change the settings.  I would like it much smaller so it only shows 4 lines of text.  That way I can increase the sizes of Row 7 and Row 9 within labelframe2

Thanks again.

**Reply by alejandroautalan** [reply]
Hello Brian,

> I have a couple of follow up questions. When I stretch the window I now get the vertical spacing between rows stretching. I do not want this. Why does that happen that way? I would like those to remain tight together at the defined 7 pady pizel size.

This happens because you have defined weight=1 for row index "all".  The all index is helpful to configure a grid where all cells will have same size, for example buttons in a calculator app.

![2022-11-01_23-45](https://user-images.githubusercontent.com/3482471/199386738-469f305c-c705-41cb-8bd3-f2df7f194da4.png)

>Also, fact I only want 2 particular rows within my labelframes stretching. Within labelframe2 I only want rows 7 and 9 to stretch (2 tkscrolledtexts). See attached UI.

Here I cleared the grid container options for the frame. And set weight=1 to rows 7 and 9 as you want. The streching of other widgets like folder labels you can manage it with the sticky option. You will see it in the updated UI file.

> Lastly, I cannot figure out how to fix the size of the tkscrolledframe_1. No matter what I do, the height is determined by the window size. You will see within the appearance, I have the height set to only 200. But nothing changes with the height regardless of how I change the settings. I would like it much smaller so it only shows 4 lines of text. That way I can increase the sizes of Row 7 and Row 9 within labelframe2

This is a tricky one. My solution to this is to use a "frame sizer". When using pack or grid manager you can't control the size
of container widgets unless you set propagate=False. So, put your tkscrolledframe inside a frame (the sizer) with the propagate=False. Now
you can set the height of the frame sizer to your preference.

![2022-11-02_00-06](https://user-images.githubusercontent.com/3482471/199387110-1bca80ad-f7b2-4b21-b69a-5d21390d11da.png)

Here is the updated UI. Review it and let me know if you have more questions.

[GUI.txt](https://github.com/alejandroautalan/pygubu-designer/files/9915865/GUI.txt)


Regards!
Alejandro A.

**Comment by 2briancox** [comment]
I am again having intermittent issues with this pygubu.builder.widgets.tkscrolledframe.

I have it inside of a frame to control its size.  

Now it intermittently stops scrolling.  Either because of a click event within an widget in the tkscrollframe, or perhaps after the appearance or disappearance of a tooltip I made. (Small Frame).

When this happens, the user can fix it by moving the mouse off of the scrollframe and then back onto it.  This is frustrating users.

My question is, what is the method/object-method to grab ahold of within the tkscrolledframe so I can control the yscroll?  My goal is to set a global bind to the whole window so that no matter where the mouse is in the window it always controls the tkscrolledframe.

I had assumed I could use tkscrolledframe.yview_scroll, but that isn't recognized.

Do you have other thoughts that may help resolve this?

**Reply by alejandroautalan** [reply]
Hello @2briancox 

> I am again having intermittent issues with this pygubu.builder.widgets.tkscrolledframe.
> I have it inside of a frame to control its size.
> Now it intermittently stops scrolling. Either because of a click event within an widget in the tkscrollframe, or perhaps after the appearance or disappearance of a tooltip I made. (Small Frame).
> When this happens, the user can fix it by moving the mouse off of the scrollframe and then back onto it. This is frustrating users.

With this I assume that you are referring to mouse-wheel scrolling.

> My question is, what is the method/object-method to grab ahold of within the tkscrolledframe so I can control the yscroll? My goal is to set a global bind to the whole window so that no matter where the mouse is in the window it always controls the tkscrolledframe.
> I had assumed I could use tkscrolledframe.yview_scroll, but that isn't recognized.
> Do you have other thoughts that may help resolve this?

The current implementation of mouse-wheel scrolling is using \<Enter\> and \<Leave\> events
to determine the current scrolling widget. I think that this could be the source of the
problems.

I started to work on a new method, dropping the usage of \<Enter\> and \<Leave\> events.

If you can, make a minimal test app to reproduce your issues.

Or it would be great if you can test your app with this new branch, and let me know if the issues
are resolved or I need to work on another approach.

    pip install git+https://github.com/alejandroautalan/pygubu@scrolledframe_issue
    
I you want to use designer with this branch use:

    pip install git+https://github.com/alejandroautalan/pygubu-designer@scrolledframe_issue


Let me know.
Alejandro A.
